x-pack/plugins/search_playground/server/utils/handle_stream_response.ts

```diff
@@ -54,15 +54,15 @@ export const handleStreamResponse = async ({
 
       push(decodedValue);
 
-      await timeout(Math.floor(Math.random() * maxTimeoutMs));
-
-      void pushStreamUpdate();
+      pushStreamUpdate();
     } catch (e) {
       logger.error(`There was an error: ${e.toString()}`);
+      end();
+      abortController.abort();
     }
   }
 
-  void pushStreamUpdate();
+  pushStreamUpdate();
```

Comment:
I've made this comment before but the whole way we handle streams still makes me uneasy, from the recursion to just handling this in a promise outside the request worries me as this is rife for potential issues.

Reply:
yeh - it has improved since the last time you saw it. It now listens to the request lifecycle and aborts early when the request dies. This is similar to the approach how others take within other streaming APIs. We could be better with tests here, having an integration test to simulate these scenarios.

It does smell theres a better way of converting the current node stream returned by langchain into a web readable stream vs the "handcoded" approach of passing through. I will have another look at this again to see if we can reduce the handcoded transformation.

Reply:
creating an issue again to look into this.